File: problems/505_the_maze_ii.py
'''
There is a ball in a maze with empty spaces (represented as 0) and walls (represented as 1). The ball can go through the empty spaces by rolling up, down, left or right, but it won't stop rolling until hitting a wall. When the ball stops, it could choose the next direction.

Given the m x n maze, the ball's start position and the destination, where start = [startrow, startcol] and destination = [destinationrow, destinationcol], return the shortest distance for the ball to stop at the destination. If the ball cannot stop at destination, return -1.

The distance is the number of empty spaces traveled by the ball from the start position (excluded) to the destination (included).

You may assume that the borders of the maze are all walls (see examples).



Example 1:


Input: maze = [[0,0,1,0,0],[0,0,0,0,0],[0,0,0,1,0],[1,1,0,1,1],[0,0,0,0,0]], start = [0,4], destination = [4,4]
Output: 12
Explanation: One possible way is : left -> down -> left -> down -> right -> down -> right.
The length of the path is 1 + 1 + 3 + 1 + 2 + 2 + 2 = 12.
Example 2:


Input: maze = [[0,0,1,0,0],[0,0,0,0,0],[0,0,0,1,0],[1,1,0,1,1],[0,0,0,0,0]], start = [0,4], destination = [3,2]
Output: -1
Explanation: There is no way for the ball to stop at the destination. Notice that you can pass through the destination but you cannot stop there.
Example 3:

Input: maze = [[0,0,0,0,0],[1,1,0,0,1],[0,0,0,0,0],[0,1,0,0,1],[0,1,0,0,0]], start = [4,3], destination = [0,1]
Output: -1


Constraints:

m == maze.length
n == maze[i].length
1 <= m, n <= 100
maze[i][j] is 0 or 1.
start.length == 2
destination.length == 2
0 <= startrow, destinationrow < m
0 <= startcol, destinationcol < n
Both the ball and the destination exist in an empty space, and they will not be in the same position initially.
The maze contains at least 2 empty spaces.
'''

class Solution:
    def shortestDistance(self, maze: List[List[int]], start: List[int], destination: List[int]) -> int:
        m, n = len(maze), len(maze[0])
        destination = tuple(destination)
        start = tuple(start)
        h = [(0, start)]
        visited = set()
        while h:
            d, v = heappop(h)
            if v in visited:
                continue
            if v == destination:
                return d
            visited.add(v)
            i,j = v
            for di, dj in [[-1,0],[1,0],[0,-1],[0,1]]:
                for k in range(max(m,n)):
                    if 0<=i+k*di<m and 0<=j+k*dj<n:
                        if maze[i+k*di][j+k*dj] == 0:
                            nxti, nxtj, nxtk = i+k*di, j+k*dj, k
                        else:
                            break
                    else:
                        break
                if (nxti, nxtj) != (i,j):
                    heappush(h, (d+nxtk,(nxti,nxtj)))
        return -1

File: problems/505_the_maze_ii.rb
=begin
There is a ball in a maze with empty spaces and walls. The ball can go through empty spaces by rolling up, down, left or right, but it won't stop rolling until hitting a wall. When the ball stops, it could choose the next direction.

Given the ball's start position, the destination and the maze, find the shortest distance for the ball to stop at the destination. The distance is defined by the number of empty spaces traveled by the ball from the start position (excluded) to the destination (included). If the ball cannot stop at the destination, return -1.

The maze is represented by a binary 2D array. 1 means the wall and 0 means the empty space. You may assume that the borders of the maze are all walls. The start and destination coordinates are represented by row and column indexes.

 

Example 1:

Input 1: a maze represented by a 2D array

0 0 1 0 0
0 0 0 0 0
0 0 0 1 0
1 1 0 1 1
0 0 0 0 0

Input 2: start coordinate (rowStart, colStart) = (0, 4)
Input 3: destination coordinate (rowDest, colDest) = (4, 4)

Output: 12

Explanation: One shortest way is : left -> down -> left -> down -> right -> down -> right.
             The total distance is 1 + 1 + 3 + 1 + 2 + 2 + 2 = 12.

Example 2:

Input 1: a maze represented by a 2D array

0 0 1 0 0
0 0 0 0 0
0 0 0 1 0
1 1 0 1 1
0 0 0 0 0

Input 2: start coordinate (rowStart, colStart) = (0, 4)
Input 3: destination coordinate (rowDest, colDest) = (3, 2)

Output: -1

Explanation: There is no way for the ball to stop at the destination.

 

Note:

There is only one ball and one destination in the maze.
Both the ball and the destination exist on an empty space, and they will not be at the same position initially.
The given maze does not contain border (like the red rectangle in the example pictures), but you could assume the border of the maze are all walls.
The maze contains at least 2 empty spaces, and both the width and height of the maze won't exceed 100.
=end

# @param {Integer[][]} maze
# @param {Integer[]} start
# @param {Integer[]} destination
# @return {Integer}
def shortest_distance(maze, start, destination)
    distance = Array.new(maze.length){Array.new(maze[0].length){Float::INFINITY}}
    dirs = [[0,1],[0,-1],[-1,0],[1,0]]
    distance[start[0]][start[1]] = 0
    queue = [start]
    while !queue.empty?
        r,c = queue.shift
        dirs.each do |dr,dc|
            nr = r + dr
            nc = c + dc
            count = 0
            while nr >= 0 && nc >= 0 && nr < maze.length && nc < maze[0].length && 
                  maze[nr][nc] == 0
                nr += dr
                nc += dc
                count += 1
            end
            nr -= dr
            nc -= dc
            if distance[r][c] + count < distance[nr][nc]
                distance[nr][nc] = distance[r][c] + count
                queue << [nr,nc]
            end
        end
    end
                
    if distance[destination[0]][destination[1]] == Float::INFINITY
        return -1
    else
        return distance[destination[0]][destination[1]]
    end
end

